### ISSUE TITLE
vSphere IPI: Bootstrap fails to complete

### ISSUE THREAD

**Issue by DizzyThermal** [openshift/okd#290]
**Describe the bug**

Tried deploying an IPI install of OKD 4.5 (latest/GA) to VMware vSphere 6.7, but bootstrapping fails to complete.

**Version**

vCenter Server Version: 6.7.0.43000 build: 15976714
openshift-install Version: 4.5.0-0.okd-2020-07-29-070316
oc Client Version: 4.5.0-0.okd-2020-07-29-070316

This is a vSphere IPI install. I haven't provisioned any other VMs other than the DNS node running BIND.

**How reproducible**

* Create DNS entries for API and Ingress:

```
api.okd4.idtus.com.             IN  A   10.4.2.71
*.apps.okd4.idtus.com.  300     IN  A   10.4.2.72
```

* Verify DNS with nslookup:

```
[admin@helper ~]$ nslookup api.okd4.idtus.com
Server:         10.14.4.1
Address:        10.14.4.1#53

Name:   api.okd4.idtus.com
Address: 10.4.2.71

[admin@helper ~]$ nslookup blah.apps.okd4.idtus.com
Server:         10.14.4.1
Address:        10.14.4.1#53

Name:   blah.apps.okd4.idtus.com
Address: 10.4.2.72
```

* Create Install Config:

```
[admin@helper ~]$ openshift-install create install-config
? SSH Public Key /home/admin/.ssh/id_rsa.pub
? Platform vsphere
? vCenter vtbvsphere01.corp.idtus.com
? Username okd4@vsphere.local
? Password [? for help] **********
INFO Connecting to vCenter vtbvsphere01.corp.idtus.com 
INFO Defaulting to only available datacenter: FRLAB 
INFO Defaulting to only available cluster: P2C    
? Default Datastore P2C-INFRAS
? Network VM Network
? Virtual IP Address for API 10.4.2.71
? Virtual IP Address for Ingress 10.4.2.72
? Base Domain idtus.com
? Cluster Name okd4
? Pull Secret [? for help] **********************************
INFO Obtaining RHCOS image file from 'https://builds.coreos.fedoraproject.org/prod/streams/stable/builds/32.20200629.3.0/x86_64/fedora-coreos-32.20200629.3.0-vmware.x86_64.ova?sha256=172f299a3e28be360740ff437a5ea9bfc246f52ea8f313d4138c5d16fd4b11e1' 
INFO The file was found in cache: /home/admin/.cache/openshift-installer/image_cache/062bfe3785d26fa220e2e6e72d1b3562. Reusing... 
INFO Creating infrastructure resources...         
INFO Waiting up to 20m0s for the Kubernetes API at https://api.okd4.idtus.com:6443... 
INFO API v1.18.3 up                               
INFO Waiting up to 40m0s for bootstrapping to complete...
0805 14:54:06.267708    2194 reflector.go:307] k8s.io/client-go/tools/watch/informerwatcher.go:146: Failed to watch *v1.ConfigMap: Get https://api.okd4.idtus.com:6443/api/v1/namespaces/kube-system/configmaps?allowWatchBookmarks=true&fieldSelector=metadata.name%3Dbootstrap&resourceVersion=3709&timeoutSeconds=475&watch=true: dial tcp 10.4.2.71:6443: connect: connection refused
E0805 14:54:07.270835    2194 reflector.go:307] k8s.io/client-go/tools/watch/informerwatcher.go:146: Failed to watch *v1.ConfigMap: Get https://api.okd4.idtus.com:6443/api/v1/namespaces/kube-system/configmaps?allowWatchBookmarks=true&fieldSelector=metadata.name%3Dbootstrap&resourceVersion=3709&timeoutSeconds=586&watch=true: dial tcp 10.4.2.71:6443: connect: connection refused
ERROR [failed to lookup master.0 ipv4 address: Post https://vtbvsphere01.corp.idtus.com/sdk: context deadline exceeded, failed to lookup master.1 ipv4 address: Post https://vtbvsphere01.corp.idtus.com/sdk: context deadline exceeded, failed to lookup master.2 ipv4 address: Post https://vtbvsphere01.corp.idtus.com/sdk: context deadline exceeded] 
INFO Pulling debug logs from the bootstrap machine 
INFO Bootstrap gather logs captured here "/home/admin/log-bundle-20200805151617.tar.gz" 
FATAL Bootstrap failed to complete: failed to wait for bootstrapping to complete: timed out waiting for the condition

```

I've tried this process more than once with the GA and latest (4.5.0-0.okd-2020-07-29-070316) build - the result was the same.

Are load balancers expected to be deployed to the provided API and Ingress VIPs during the installation, or does the IPI install do that as well?

I've tried configuring two HAProxy loadbalancers on two separate RHEL 7 VMs:

* 10.4.2.71 (API)
* 10.4.2.72 (Ingress)

Each VM's haproxy.cfg pointed to 3 non-existent VMs:

* 10.4.2.71 - API to: 10.4.2.73-75
* 10.4.2.72 - Ingress to: 10.4.2.76-78

Bootstrapping still failed in this test as well. I'm not quiet sure if that configuration is right - is there documentation on how the load balancers are expected to be configured for OKD4?

I wasn't sure how similar this was to [issues/276](https://github.com/openshift/okd/issues/276), so I wanted to provide my steps to reproduce and log bundle.

**Log bundle**
[log-bundle-20200805151617.tar.gz](https://github.com/openshift/okd/files/5030836/log-bundle-20200805151617.tar.gz)

Thanks in advance!

**Comment by vrutkovs** [comment]
Looks very similar to #276 - bootstrap didn't get proper hostname assigned. Would you mind giving one of the latest nightlies a try?

**Comment by DizzyThermal** [comment]
Okay, trying with nightly: `release:4.5.0-0.okd-2020-08-06-102301`

Running:
```
oc adm release extract --tools registry.svc.ci.openshift.org/origin/release:4.5.0-0.okd-2020-08-06-102301
```

Then repeating the steps in my original post to perform the vSphere IPI install

Will post results shortly. Thanks @vrutkovs !

**Comment by DizzyThermal** [comment]
[release:4.5.0-0.okd-2020-08-06-102301](https://origin-release.apps.ci.l2s4.p1.openshiftapps.com/releasestream/4.5.0-0.okd/release/4.5.0-0.okd-2020-08-06-102301) also failed - here is what I did:

```
[admin@helper ~]$ oc adm release extract --tools registry.svc.ci.openshift.org/origin/release:4.5.0-0.okd-2020-08-06-102301
[admin@helper ~]$ openshift-install create cluster
? SSH Public Key /home/admin/.ssh/id_rsa.pub
? Platform vsphere
? vCenter vtbvsphere01.corp.idtus.com
? Username okd4@vsphere.local
? Password [? for help] **********
INFO Connecting to vCenter vtbvsphere01.corp.idtus.com 
INFO Defaulting to only available datacenter: FRLAB 
INFO Defaulting to only available cluster: P2C    
? Default Datastore P2C-INFRAS
? Network VM Network
? Virtual IP Address for API 10.4.2.71
? Virtual IP Address for Ingress 10.4.2.72
? Base Domain idtus.com
? Cluster Name okd4
? Pull Secret [? for help] **********************************
INFO Obtaining RHCOS image file from 'https://builds.coreos.fedoraproject.org/prod/streams/stable/builds/32.20200629.3.0/x86_64/fedora-coreos-32.20200629.3.0-vmware.x86_64.ova?sha256=172f299a3e28be360740ff437a5ea9bfc246f52ea8f313d4138c5d16fd4b11e1' 
INFO The file was found in cache: /home/admin/.cache/openshift-installer/image_cache/062bfe3785d26fa220e2e6e72d1b3562. Reusing... 
INFO Creating infrastructure resources...         
INFO Waiting up to 20m0s for the Kubernetes API at https://api.okd4.idtus.com:6443... 
ERROR Attempted to gather ClusterOperator status after installation failure: listing ClusterOperator objects: Get https://api.okd4.idtus.com:6443/apis/config.openshift.io/v1/clusteroperators: dial tcp 10.4.2.71:6443: connect: no route to host 
ERROR Attempted to gather debug logs after installation failure: failed to get bootstrap and control plane host addresses from "terraform.tfstate": failed to lookup bootstrap ipv4 address: Post https://vtbvsphere01.corp.idtus.com/sdk: context deadline exceeded 
FATAL Bootstrap failed to complete: failed waiting for Kubernetes API: Get https://api.okd4.idtus.com:6443/version?timeout=32s: dial tcp 10.4.2.71:6443: connect: no route to host
```

Grabbed the log-bundle using: `[admin@helper ~]$ openshift-install gather bootstrap`

**Log Bundle**:

[log-bundle-20200805151617.tar.gz](https://github.com/openshift/okd/files/5035370/log-bundle-20200805151617.tar.gz)

**Comment by DizzyThermal** [comment]
Trying [4.5.0-0.okd-2020-08-07-103751](oc adm release extract --tools registry.svc.ci.openshift.org/origin/release:4.5.0-0.okd-2020-08-07-103751) - will post results shortly.

**Comment by DizzyThermal** [comment]
The [4.5.0-0.okd-2020-08-07-103751](https://origin-release.apps.ci.l2s4.p1.openshiftapps.com/releasestream/4.5.0-0.okd/release/4.5.0-0.okd-2020-08-07-103751) release page says at the top that this command:

`oc adm release extract --tools registry.svc.ci.openshift.org/origin/release:4.5.0-0.okd-2020-08-07-103751`

downloads the client and installer.

The `oc` and `openshift-install` binaries under `/usr/local/bin` are still the ones I manually downloaded from: https://github.com/openshift/okd/releases/tag/4.5.0-0.okd-2020-07-29-070316

Where is the `oc adm release extract --tools` command downloading the installer and client to, I'd like to move it over the 2020-07-29 release binaries.

Thanks!

**Comment by DizzyThermal** [comment]
Gotcha, extracts to cwd. _Actually_ testing [4.5.0-0.okd-2020-08-07-103751](https://origin-release.apps.ci.l2s4.p1.openshiftapps.com/releasestream/4.5.0-0.okd/release/4.5.0-0.okd-2020-08-07-103751)

Will post results shortly

**Comment by DizzyThermal** [comment]
Okay, so I ran with the above version's client/installer binaries updated:

```
[admin@helper ~]$ oc version                                                                                                                                                                                       
Client Version: 4.5.0-0.okd-2020-08-07-103751                                                                                                                                                                      
[admin@helper ~]$ openshift-install version                                                                                                                                                                        
openshift-install 4.5.0-0.okd-2020-08-07-103751
built from commit 699277bb61706731d687b9e40700ebf4630b0851
release image registry.svc.ci.openshift.org/origin/release@sha256:4d72ca8ae73bb4b6596464485bce191d8dbef5f5d71f5c91d49dd89fc1dc7786
```

Running with `log-level=debug` once it got to waiting for the Kubernetes API:

```
INFO Waiting up to 20m0s for the Kubernetes API at https://api.okd4.idtus.com:6443... 
DEBUG Still waiting for the Kubernetes API: Get https://api.okd4.idtus.com:6443/version?timeout=32s: dial tcp 10.4.2.71:6443: connect: no route to host 
DEBUG Still waiting for the Kubernetes API: Get https://api.okd4.idtus.com:6443/version?timeout=32s: dial tcp 10.4.2.71:6443: connect: no route to host 
DEBUG Still waiting for the Kubernetes API: Get https://api.okd4.idtus.com:6443/version?timeout=32s: dial tcp 10.4.2.71:6443: connect: no route to host 
DEBUG Still waiting for the Kubernetes API: Get https://api.okd4.idtus.com:6443/version?timeout=32s: dial tcp 10.4.2.71:6443: connect: no route to host 
DEBUG Still waiting for the Kubernetes API: Get https://api.okd4.idtus.com:6443/version?timeout=32s: dial tcp 10.4.2.71:6443: connect: no route to host 
DEBUG Still waiting for the Kubernetes API: Get https://api.okd4.idtus.com:6443/version?timeout=32s: dial tcp 10.4.2.71:6443: connect: no route to host 
DEBUG Still waiting for the Kubernetes API: Get https://api.okd4.idtus.com:6443/version?timeout=32s: dial tcp 10.4.2.71:6443: connect: no route to host 
DEBUG Still waiting for the Kubernetes API: Get https://api.okd4.idtus.com:6443/version?timeout=32s: dial tcp 10.4.2.71:6443: connect: no route to host 
DEBUG Still waiting for the Kubernetes API: Get https://api.okd4.idtus.com:6443/version?timeout=32s: dial tcp 10.4.2.71:6443: connect: no route to host 
DEBUG Still waiting for the Kubernetes API: Get https://api.okd4.idtus.com:6443/version?timeout=32s: dial tcp 10.4.2.71:6443: connect: no route to host 
DEBUG Still waiting for the Kubernetes API: Get https://api.okd4.idtus.com:6443/version?timeout=32s: dial tcp 10.4.2.71:6443: connect: no route to host 
DEBUG Still waiting for the Kubernetes API: Get https://api.okd4.idtus.com:6443/version?timeout=32s: dial tcp 10.4.2.71:6443: connect: no route to host 
DEBUG Still waiting for the Kubernetes API: Get https://api.okd4.idtus.com:6443/version?timeout=32s: dial tcp 10.4.2.71:6443: connect: no route to host 
DEBUG Still waiting for the Kubernetes API: Get https://api.okd4.idtus.com:6443/version?timeout=32s: dial tcp 10.4.2.71:6443: connect: no route to host 
DEBUG Still waiting for the Kubernetes API: Get https://api.okd4.idtus.com:6443/version?timeout=32s: dial tcp 10.4.2.71:6443: connect: no route to host 
DEBUG Still waiting for the Kubernetes API: Get https://api.okd4.idtus.com:6443/version?timeout=32s: dial tcp 10.4.2.71:6443: connect: no route to host 
DEBUG Still waiting for the Kubernetes API: Get https://api.okd4.idtus.com:6443/version?timeout=32s: dial tcp 10.4.2.71:6443: connect: no route to host 
DEBUG Still waiting for the Kubernetes API: Get https://api.okd4.idtus.com:6443/version?timeout=32s: dial tcp 10.4.2.71:6443: connect: no route to host 
DEBUG Still waiting for the Kubernetes API: Get https://api.okd4.idtus.com:6443/version?timeout=32s: dial tcp 10.4.2.71:6443: connect: no route to host 
DEBUG Still waiting for the Kubernetes API: Get https://api.okd4.idtus.com:6443/version?timeout=32s: dial tcp 10.4.2.71:6443: connect: no route to host 
DEBUG Still waiting for the Kubernetes API: Get https://api.okd4.idtus.com:6443/version?timeout=32s: dial tcp 10.4.2.71:6443: connect: no route to host 
DEBUG Still waiting for the Kubernetes API: Get https://api.okd4.idtus.com:6443/version?timeout=32s: dial tcp 10.4.2.71:6443: connect: no route to host 
DEBUG Still waiting for the Kubernetes API: Get https://api.okd4.idtus.com:6443/version?timeout=32s: dial tcp 10.4.2.71:6443: connect: no route to host 
DEBUG Still waiting for the Kubernetes API: Get https://api.okd4.idtus.com:6443/version?timeout=32s: dial tcp 10.4.2.71:6443: connect: no route to host 
DEBUG Still waiting for the Kubernetes API: Get https://api.okd4.idtus.com:6443/version?timeout=32s: dial tcp 10.4.2.71:6443: connect: no route to host 
DEBUG Still waiting for the Kubernetes API: Get https://api.okd4.idtus.com:6443/version?timeout=32s: dial tcp 10.4.2.71:6443: connect: no route to host 
ERROR Attempted to gather ClusterOperator status after installation failure: listing ClusterOperator objects: Get https://api.okd4.idtus.com:6443/apis/config.openshift.io/v1/clusteroperators: dial tcp 10.4.2.71:6443: connect: no route to host 
DEBUG Fetching Bootstrap SSH Key Pair...           
DEBUG Loading Bootstrap SSH Key Pair...            
DEBUG Using Bootstrap SSH Key Pair loaded from state file 
DEBUG Reusing previously-fetched Bootstrap SSH Key Pair 
DEBUG Fetching Install Config...                   
DEBUG Loading Install Config...                    
DEBUG   Loading SSH Key...                         
DEBUG   Loading Base Domain...                     
DEBUG     Loading Platform...                      
DEBUG   Loading Cluster Name...                    
DEBUG     Loading Base Domain...                   
DEBUG     Loading Platform...                      
DEBUG   Loading Pull Secret...                     
DEBUG   Loading Platform...                        
DEBUG Using Install Config loaded from state file  
DEBUG Reusing previously-fetched Install Config    
ERROR Attempted to gather debug logs after installation failure: failed to get bootstrap and control plane host addresses from "terraform.tfstate": failed to lookup bootstrap ipv4 address: Post https://vtbvsphere01.corp.idtus.com/sdk: context deadline exceeded 
FATAL Bootstrap failed to complete: failed waiting for Kubernetes API: Get https://api.okd4.idtus.com:6443/version?timeout=32s: dial tcp 10.4.2.71:6443: connect: no route to host
```

I tried to use: `openshift-install gather bootstrap`, to grab the log-bundle, but it failed to access the terraform VMs:

```
[admin@helper ~]$ openshift-install gather bootstrap
FATAL failed to get bootstrap and control plane host addresses from "terraform.tfstate": failed to lookup bootstrap ipv4 address: Post https://vtbvsphere01.corp.idtus.com/sdk: context deadline exceeded 
```

All 3 master VMs are powered on and looping with the following:

`[ 2075.270779] ignition[501]: GET error: Get "https://10.4.2.71:22623/config/master": dial tcp 10.4.2.71:22626: connect: network is unreachable`

It seems to still not be setting the hostname correctly? Is there another nightly I could try to use?

Thanks again!

**Comment by rarguello** [comment]
Does your DNS server resolve a reverse zone for your ip addresses?
I had to create the reverse zone for the installer to finish.

**Comment by DizzyThermal** [comment]
@rarguello 

My DNS Server has the following forward zone entries:

```
api.okd4.idtus.com.             IN  A   10.4.2.71
*.apps.okd4.idtus.com.  300     IN  A   10.4.2.72
```

...and a reverse zone for the API address:
```
71              IN  PTR         api.okd4.idtus.com.
```

If I try to add one for Ingress (wildcard), BIND will not start (example below):

```
72              IN  PTR         *.apps.okd4.idtus.com.
```

The BIND error is:

```
bash[20345]: dns_rdata_fromtext: 2.4.10.rev.idtus.com:16: near '*.apps.okd4.idtus.com.': bad name (check-names)
bash[20345]: zone 2.4.10.in-addr.arpa/IN: loading from master file 2.4.10.rev.idtus.com failed: bad name (check-names)
```

Maybe not having the wildcard reverse zone is what I'm missing?

**Comment by rarguello** [comment]
You need a reverse entry for every master and worker.
I don't think you can have a reverse for a wildcard, you don't really need that.

**Comment by nate-duke** [comment]
Is it true that we need a reverse record prior to cluster installation for every master and worker? That's not mentioned in the DNS section of the installation documentation. The docs only mention a *forward* for the wildcard and the api address.

**Comment by vrutkovs** [comment]
You don't need it, unless your DHCP is misconfigured and sets unexpected IP to the nodes. Configuring this is not an OKD bug really

**Comment by bogd** [comment]
@vrutkovs - please clarify. What do "unexpected IP" or "DHCP misconfigured" mean for you?

I am asking because I am facing a similar issue (bootstrap not completing), and this thread actually gives conflicting information. Is a reverse DNS entry required or not? If it is, when exactly is it required? And are there any restrictions for the PTR names (i.e. do they need to match the forward entries)?

Thank you!

**Comment by vrutkovs** [comment]
Once your masters get an IP its not expected to change (this causes issues with certificates). So your DHCP setup should issue a lease for masters in IPI and have it persisted

**Comment by igpix** [comment]
@DizzyThermal  have you been able to solve the issue? same problem here

**Comment by DizzyThermal** [comment]
@igpix - no, I decided to take the UPI approach following Craig Robinson's tutorial/guide for OKD 4.5.

**Comment by igpix** [comment]
@DizzyThermal At the end I got it working. IPI installation on vsphere can require more time then the one available with 'create cluster' that ends for time out.  I had to ran 'wait-for bootstrap-complete' to complete the bootstrap and then 'wait-for install-complete'. Others hints: make sure to have *all* available hw requirements and configure dhcp server with correct lease time (increasing it as much as possible).
